File: 프로그래머스/lv3/43238. 입국심사/입국심사.py
def solve(times, t):
    res = 0
    for i in range(len(times)):
        res += (t // times[i])
    return res

def solution(n, times):
    l = 0
    h = n * max(times)
    
    answer = float('inf')
    while l <= h:
        mid = (l + h) // 2
        person_count = solve(times, mid)
        if person_count < n:
            l = mid + 1
        else:
            answer = min(answer, mid)
            h = mid - 1

    return answer

File: 프로그래머스/lv3/43238. 입국심사/입국심사.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public long calPerson(long t, int[] times){
        long res = 0;
        for(int i = 0; i < times.length; ++i)
            res += (long)(t / times[i]);
        return res;
    }
    public long solution(int n, int[] times) {
        long answer = (long)n * Arrays.stream(times).max().getAsInt() + 1;
        long s = 0;
        long e = (long)n * Arrays.stream(times).max().getAsInt() + 1;
        
        while(s <= e){
            long mid = (long)((e + s) / 2);
            long people = calPerson(mid, times);
            if(people < n){
                s = mid + 1;
            }else{
                answer = Math.min(answer, mid);
                e = mid - 1;
            }
            
        }
        return answer;
    }
}